hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodiePositionBasedFileGroupRecordBuffer.java

```diff
@@ -123,46 +142,42 @@ public void processDataBlock(HoodieDataBlock dataBlock, Option<KeySpec> keySpecO
     }
   }
 
-  @Override
-  public void processNextDataRecord(T record, Map<String, Object> metadata, Serializable recordPosition) throws IOException {
-    Pair<Option<T>, Map<String, Object>> existingRecordMetadataPair = records.get(recordPosition);
-    Option<Pair<T, Map<String, Object>>> mergedRecordAndMetadata =
-        doProcessNextDataRecord(record, metadata, existingRecordMetadataPair);
-    if (mergedRecordAndMetadata.isPresent()) {
-      records.put(recordPosition, Pair.of(
-          Option.ofNullable(readerContext.seal(mergedRecordAndMetadata.get().getLeft())),
-          mergedRecordAndMetadata.get().getRight()));
+  private void fallbackToKeyBasedBuffer() {
+    readerContext.setShouldMergeUseRecordPosition(false);
+    //need to make a copy of the keys to avoid concurrent modification exception
+    ArrayList<Serializable> positions = new ArrayList<>(records.keySet());
```

Comment:
rename `positions` to `recordKeys` and `position` to `recordKey`, or something like that.. because this is not position right.

Reply:
No, those are positions. The map is recordpositon->record. After we fallback it becomes recordkey->record